src/main.cpp

```diff
@@ -387,25 +389,42 @@ bool MarkBlockAsReceived(const uint256& hash) {
 }
 
 // Requires cs_main.
-void MarkBlockAsInFlight(NodeId nodeid, const uint256& hash, const Consensus::Params& consensusParams, CBlockIndex *pindex = NULL) {
+// When pit is set it is assumed a PartiallyDownloadedBlock will be used
+// pit is only set if a new PartiallyDownloadedBlock was created
+// pit will only be valid as long as the same cs_main lock is being held
```

Comment:
That's pretty unintuitive behaviour, and I don't think it's necessary to put that much decision logic in this function. I think it could just return a `const QueuedBlock*` always, and then the caller can check whether its partialBlock field was set?